### ISSUE TITLE
no-unused-vars is incorrectly reported

### ISSUE THREAD

**Issue by zacleids** [eslint/eslint#9717]
**Tell us about your environment**

* **ESLint Version:** v4.13.1
* **Node Version:** v6.9.5
* **npm Version:** 3.10.10

**What parser (default, Babel-ESLint, etc.) are you using?**
default
**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "env": {
    "es6": true,
    "node": true
  },
  "parserOptions": {
    "ecmaVersion": 6
  },
  "rules": {
    "no-unused-vars": ["error", {"vars": "all", "args": "none"}],
    "no-redeclare": "error"
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const fs = require('fs');
const folder = '.';
const filenames = fs.readdirSync(folder);

for (const fileName of filenames) {
  if (!(fileName.startsWith('v1') && !fileName.endsWith('.config.json'))) {
    continue;
  }

  const fileName = this.args.file;
  const data = JSON.parse(fs.readFileSync(`${folder}/${fileName}`));

  //processData
  const newData = data.valueOf();

  const newFile = `${folder}/${fileName.replace('processed', 'completed')}`;
  fs.writeFileSync(newFile, newData, { encoding: 'utf8', mode: '0600' });
}
```

```bash
eslint file.js
```

**What did you expect to happen?**
to get a no-redeclare error from the variable fileName on line 10
**What actually happened? Please include the actual, raw output from ESLint.**
```
  5:12  error  'fileName' is assigned a value but never used  no-unused-vars

✖ 1 problem (1 error, 0 warnings)
```
If you remove `no-redeclare` from the configuration file, the same output is produced.

**Comment by not-an-aardvark** [comment]
Hi,

Thanks for the report. This is actually working as intended; the inside of a `for` loop has a different scope than the header, so `fileName` variable on line 10 is shadowing the `fileName` variable on line 5. `no-redeclare` reports multiple declarations of the *same* variable; it doesn't report an error here because two separate variables are being used in different scopes.

`no-unused-vars` is reporting an error because the `fileName` variable declared on line 5 is never used. (Although it might seem that the outer `fileName` variable is used on line 6, the reference to `fileName` on line 6 actually refers to the inner `fileName` variable.) Your code will probably throw a runtime error because the inner `fileName` variable has not yet been defined on line 6.

To avoid confusing cases like this, you might be interested in the [`no-shadow`](https://eslint.org/docs/rules/no-shadow).

**Comment by platinumazure** [comment]
Closing as the question has been answered. Please feel free to comment here or stop by our [Gitter chat](https://gitter.im/eslint/eslint) if you have any further questions. Thanks!
